### PR TITLE
raftstore-v2: support to make protection when disk full.

### LINKED ISSUE #15170: [Dynamic Regions] Disk full protection

### PR DESCRIPTION
**Pull request by LykxSassinator**
### What is changed and how it works?

Issue Number: Close #15170 

What's Changed:

```commit-message
This pr is used to protect `raftstore-v2` when disk full. And all checking and
validation is transplant from `raftstore`.
```

All compatible implementations in `raftstore-v2` are transplanted from `raftstore`, and the difference between these two part can be reviewed as followings show:
DiskFullOpt

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List
 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
Raftstore-v2 can percept `DiskFull` exceptions and refuse to execute `Write` operations when disk full.
```

### CODE REVIEW

**Comment by LykxSassinator** [comment]
/cc @overvenus @tonyxuqqi 
PTAL

File: components/raftstore-v2/src/operation/command/admin/mod.rs [hunk 1]
```diff
@@ -134,9 +147,39 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
             ch.report_error(resp);
             return;
         }
+        // Prepare Merge need to be broadcast to as many as followers when disk full.
+        let is_merge_cmd =
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/operation/command/admin/mod.rs:151]: please wrap this new block into a separate function.

File: components/raftstore-v2/src/operation/command/admin/split.rs [hunk 2]
```diff
@@ -365,6 +374,13 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
             info!(self.logger, "not leader, skip.");
             return;
         }
+        // Check whether the admin request can be proposed when disk full.
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/operation/command/admin/split.rs:377]: why it's repeated here.

File: components/raftstore-v2/src/operation/life.rs [hunk 3]
```diff
@@ -126,18 +132,23 @@ pub struct AbnormalPeerContext {
...
     pub fn is_empty(&self) -> bool {
-        self.pending_peers.is_empty() && self.down_peers.is_empty()
+        self.pending_peers.is_empty() && self.down_peers.is_empty() /* && self.disk_full_peers.is_empty() */
     }
 
     #[inline]
     pub fn reset(&mut self) {
         self.pending_peers.clear();
         self.down_peers.clear();
+        // self.disk_full_peers.clear();
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/operation/life.rs:151]: is it needed?

**Review by tonyxuqqi — APPROVED** [review]
overall LGTM

File: components/raftstore-v2/src/operation/life.rs (hunk 3 above)
Comment (LykxSassinator) [reply components/raftstore-v2/src/operation/life.rs:151]: Missed it. It's redundant here, as the disk_full_peers will be refreshed at each tick if the disk_usage is updated.

File: components/raftstore-v2/src/operation/command/admin/split.rs (hunk 2 above)
Comment (LykxSassinator) [reply components/raftstore-v2/src/operation/command/admin/split.rs:377]: Maybe u mean the check in RequestSplit & RequestHalfSplit?
image
 
The checks on disk_full  are both needed.

**Review by overvenus — APPROVED** [review]

**Pull request merged**